Add Accession Export via Clipboard (BS5)

danloveg opened the pull request:
Closes #1901 

This is a ~~mostly~~ complete implementation of exporting accessions via the clipboard in the `arDominionBS5Plugin` theme.

Changes include:

- Revamp accession browse results using a new `_searchResult` partial template.
- Add a clipboard button to accession pages.
- Add a new UI label for accessions.
- Conditionally show accessions in clipboard whether the authenticated user has permission to view accessions.
- Add a new job to export accessions as CSV from the frontend.

Here is an image of accessions on the clipboard:

![accession_clipboard](https://github.com/user-attachments/assets/4bb6f5a7-d329-4e45-9b99-b68b25e5d9a5)

Things currently missing/needing work:

- [x] There's no "Export CSV" from the Accession browse page.
- [x] If you add an accession to the clipboard, then log out, there's a phantom item on the clipboard. This is because un-authenticated users cannot see accession records. I'm not sure how to handle this.

danloveg committed:
Add accessions to clipboard.js

danloveg committed:
Add accessions to clipboard.js

danloveg committed:
Add clipboard component to accession page

danloveg committed:
Add accession export job files (untested)

danloveg committed:
Add accession to clipboard module

danloveg committed:
Add UI label for accessions

danloveg committed:
Fix clipboard search filters for accessions

danloveg committed:
Render accessions on clipboard view page

danloveg committed:
Add accession label to settings.yml

danloveg committed:
Revert changes to non-BS5 template

danloveg committed:
Hide accession export from users w/o permission

danloveg committed:
Hide accession export from users w/o permission

danloveg committed:
Register new job in gearman.yml

danloveg committed:
Change accession "identifier" to Accession number

danloveg committed:
Replace table with searchResult partial template

danloveg committed:
Run php-cs-fixer

danloveg committed:
Run prettier

danloveg commented:
@melaniekung let me know if you want to collaborate on this - I saw you had started doing some of this work in parallel in https://github.com/artefactual/atom/commit/3cd087eb10a6e28dd6bb707ef41b721e6701d69a

danloveg committed:
Merge branch 'qa/2.x' into dev/BS5-accession-exporter

danloveg committed:
Hide accessions when un-authenticated

danloveg committed:
Hide accessions when un-authenticated

danloveg committed:
Run prettier

danloveg committed:
Add new migration back

danloveg committed:
Add Export CSV button to accession browse page

danloveg committed:
Merge branch 'qa/2.x' into dev/BS5-accession-exporter

danloveg committed:
Fix php-cs-fixer issue

danloveg committed:
Use existing i18n strings for message

danloveg commented:
Note: This feature doesn't work without #2118.

I had a couple tasks I wanted to finish before this was ready:

> There's no "Export CSV" from the Accession browse page.

I've added this, but the button goes away when filtering the accessions in the search box. So you can only export all of the accessions using this button.

> If you add an accession to the clipboard, then log out, there's a phantom item on the clipboard.

The total number of items on the clipboard now dynamically updates when the user logs in or out.

melaniekung commented:
Hi @danloveg , if you have the capacity and can take this on, that would be really appreciated. That said, if you don't have the capacity, we can pick this back up once we have the capacity.

Since you've mentioned having a couple tasks to finish, we just wanted to give you an early FYI as we will be putting out a formal announcement soon, that we will be removing BS2 from AtoM in 2.10. We've realized that it will require a large amount of work to rebase any existing pull requests once we merge this to `qa/2.x` as this commit will include moving multiple templates and javascripts from `plugins/arDominionB5Plugin` to `apps/qubit` or `js/`. Therefore, we plan to merge/close all existing PRs in a month or so (a more exact date will come in the official announcement) before we do the BS2 removal commit.

danloveg commented:
> Hi @danloveg , if you have the capacity and can take this on, that would be really appreciated. That said, if you don't have the capacity, we can pick this back up once we have the capacity.

Thanks for checking in. I do have the capacity to work on this so let me know if you need anything else in regards to this pull request. I've cleaned it up in the past couple of days, all this needs now is a review.

> Since you've mentioned having a couple tasks to finish

The two tasks I mentioned earlier are done now! This PR is review-ready

> we just wanted to give you an early FYI as we will be putting out a formal announcement soon, that we will be removing BS2 from AtoM in 2.10. We've realized that it will require a large amount of work to rebase any existing pull requests once we merge this to `qa/2.x` as this commit will include moving multiple templates and javascripts from `plugins/arDominionB5Plugin` to `apps/qubit` or `js/`. Therefore, we plan to merge/close all existing PRs in a month or so (a more exact date will come in the official announcement) before we do the BS2 removal commit.

That makes sense. I have time this month to update any of my changes on this branch so do let me know if you need me to make any changes

danloveg committed:
Add accessioncount to settings fixture

danloveg committed:
Merge branch 'qa/2.x' into dev/BS5-accession-exporter

Move new migration from 0195 to 0196 to fix conflict.

danloveg committed:
Fix shifted columns in CSV exports (#2123)

* Fix shifting columns due to hidden fields

* Add unit tests for exportResource

danloveg committed:
Change colour by targeting id (#2129)

danloveg committed:
Unit test to create objects in transaction (#2125)

* Run unit test in a transaction

* Add docstring to withTransaction

* Convert function to re-usable TransactionTestCase class

* Fix docstring with php-cs-fixer

danloveg committed:
Logging for search:populate --slug option (#2124)

danloveg committed:
Add user to export context (#2117)

* Set user, environment in QubitFlatfileExport

* Tweak logic excluding accessionNumber

The accessionNumber is excluded if:

- The environment is not cli or worker
- AND, the user is not set, or the user is not authenticated

danloveg committed:
Run npm run format

danloveg committed:
Merge branch 'qa/2.x' into dev/BS5-accession-exporter

danloveg committed:
Add checks to prevent users w/o permission to export

melaniekung commented on apps/qubit/modules/clipboard/actions/exportAction.class.php:
```diff
@@ -166,6 +171,15 @@ public function execute($request)
             return $this->renderText(json_encode(['error' => $message]));
         }
 
+        if ('QubitAccession' === $className && !$this->context->user->hasCredential(['contributor', 'editor', 'administrator'], false)) {
```
hi @danloveg , we think that only users who have edit permissions to  be able to export accessions, and so since contributors don't have edit permissions to accessions, so they shouldn't be  able to export an accession.

danloveg replied:
That's a great point, I didn't realize they weren't able to edit accessions.

I disabled exporting accessions for contributors in https://github.com/artefactual/atom/pull/2101/commits/51584d1de3d3997b1d03fa3dfe7bdbbd6823ee0a

melaniekung commented on apps/qubit/modules/clipboard/actions/viewAction.class.php:
```diff
@@ -66,6 +66,11 @@ public function execute($request)
         ];
 
+        // Some users cannot see accessions - don't show them in that case
+        if ($this->context->user->hasCredential(['contributor', 'editor', 'administrator'], false)) {
```
same issue as mentioned earlier, only those with accession edit permissions should be able to export accesions

danloveg replied:
Done, see: https://github.com/artefactual/atom/pull/2101/commits/51584d1de3d3997b1d03fa3dfe7bdbbd6823ee0a

melaniekung commented on plugins/arDominionB5Plugin/modules/menu/templates/_clipboardMenu.php:
```diff
@@ -13,7 +13,14 @@ class="nav-link dropdown-toggle d-flex align-items-center p-0"
         'Note: clipboard items unclipped in this page will be removed from the clipboard when the page is refreshed. You can re-select them now, or reload the page to remove them completely. Using the sort or print preview buttons will also cause a page reload - so anything currently deselected will be lost!'
-    ); ?>">
+    ); ?>"
+    <?php if ($sf_user->hasCredential(['contributor', 'editor', 'administrator'], false)) { ?>
```
and again, contributors should be excluded from being able to export accessions

danloveg replied:
Done, see: https://github.com/artefactual/atom/pull/2101/commits/51584d1de3d3997b1d03fa3dfe7bdbbd6823ee0a

melaniekung commented on data/fixtures/settings.yml:
hi @danloveg , one more note - please update the value [here](https://github.com/artefactual/atom/blob/60917f43c39d8edad0b1b6664a4d0596568fdf4c/data/fixtures/settings.yml#L6) to 196 for the new migration file you've added

danloveg replied:
For sure, I've done that here: https://github.com/artefactual/atom/pull/2101/commits/44cc899be27144996d1e75fb74400dd021297e12. I'll make a note to myself that this file should be updated when adding migrations.

danloveg committed:
Prevent contributors from exporting accessions

Contributors can still view accessions, but they can't edit them. This
commit disables them from exporting accessions.

The clipboard icon and the Export CSV button are conditionally hidden in
the rendered templates as well.

danloveg committed:
Update latest migration version in settings fixture

danloveg commented:
The accession page looks like this to a contributor now, note the absence of the Export CSV button and the clipboard icons.

![contributor_accession_view](https://github.com/user-attachments/assets/2b17d926-cead-449d-8a29-df3c2387072c)

danloveg committed:
Add missing second arg (do not use "and" logic)

danloveg commented on plugins/arDominionB5Plugin/modules/accession/templates/_searchResult.php:
```diff
@@ -0,0 +1,62 @@
+          ['class' => 'h5 mb-0 text-truncate'],
+      ); ?>
+
+      <?php if ($canExportAccessions) { ?>
```
This new variable isn't defined on the page `/clipboard/view?type=accession` so these aren't showing up on the clipboard page.

danloveg replied:
Now it's working: https://github.com/artefactual/atom/pull/2101/commits/cfa152560d8f3a4f63e04631362e6e54a924426b

danloveg committed:
Define $canExportAccessions in clipboard view

danloveg committed:
Add more permission hardening

melaniekung reviewed (approved):
Looks good, thanks so much for taking this on, @danloveg !

danloveg commented:
> Looks good, thanks so much for taking this on, @danloveg !

Thanks for taking the time to review, glad I could contribute this feature!
